compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -791,40 +791,83 @@ impl<'a, 'tcx> CastCheck<'tcx> {
+
+            // trait object -> trait object? need to do additional checks
+            (Some(PointerKind::VTable(src_tty)), Some(PointerKind::VTable(dst_tty))) => {
+                match (src_tty.principal(), dst_tty.principal()) {
+                    // A<dyn Trait + Auto> -> B<dyn Trait' + Auto'>. need to make sure
+                    // - traits are the same & have the same generic arguments
+                    // - Auto' is a subset of Auto
+                    //
+                    // This is checked by checking `dyn Trait + Auto + 'erased: Unsize<dyn Trait' + Auto' + 'erased>`.
+                    (Some(_), Some(_)) => {
+                        let tcx = fcx.tcx;
+
+                        // We need to reconstruct trait object types.
+                        // `m_src` and `m_dst` won't work for us here because they will potentially
+                        // contain wrappers, which we do not care about.
+                        //
+                        // e.g. we want to allow `dyn T -> (dyn T,)`, etc.
+                        let src_obj = tcx.mk_ty_from_kind(ty::Dynamic(src_tty, tcx.lifetimes.re_erased, ty::Dyn));
+                        let dst_obj = tcx.mk_ty_from_kind(ty::Dynamic(dst_tty, tcx.lifetimes.re_erased, ty::Dyn));
+
+                        // `dyn Src: Unsize<dyn Dst>`
+                        let cause = fcx.misc(self.span);
+                        let obligation = Obligation::new(
+                            tcx,
+                            cause,
+                            fcx.param_env,
+                            ty::TraitRef::from_lang_item(
+                                tcx,
+                                LangItem::Unsize,
+                                self.span,
+                                [src_obj, dst_obj]
+                            )
+                        );
 
-        // thin -> fat? report invalid cast (don't complain about vtable kinds)
-        if expr_kind == PointerKind::Thin {
-            return Err(CastError::SizedUnsizedCast);
-        }
+                        fcx.register_predicate(obligation);
 
-        // vtable kinds must match
-        if fcx.tcx.erase_regions(cast_kind) == fcx.tcx.erase_regions(expr_kind) {
-            Ok(CastKind::PtrPtrCast)
-        } else {
-            Err(CastError::DifferingKinds)
+                        // FIXME: ideally we'd maybe add a flag here, so that borrowck knows that
+                        //        it needs to borrowck this ptr cast. this is made annoying by the
+                        //        fact that `thir` does not have `CastKind` and mir restores it
+                        //        from types.
+                        Ok(CastKind::PtrPtrCast)
+                    }
+
+                    // dyn Auto -> dyn Auto'? ok.
+                    (None, None)
+                    // dyn Trait -> dyn Auto? ok.
+                    | (Some(_), None)=> Ok(CastKind::PtrPtrCast),
```

Comment:
Just to point it out: We currently reject these kinds of casts. But allowing them seems fine to me.

Reply:
I think I'll change this to the current behavior for now, to keep the decision space for this PR smaller.

Reply:
I don't think this happened?

Reply:
This happened, you reviewed an old commit ^^'